File: framework/wcf/Extensibility/Transactions/TransactionMessagePropertyUdpTransport/VB/UdpTransport/UdpOutputChannel.vb
'  Copyright (c) Microsoft Corporation. All rights reserved.

Imports System.Globalization
Imports System.Net
Imports System.Net.Sockets
Imports System.ServiceModel
Imports System.ServiceModel.Channels

Namespace Microsoft.ServiceModel.Samples
	''' <summary>
	''' IOutputChannel implementation for Udp.
	''' </summary>
	Friend Class UdpOutputChannel
		Inherits ChannelBase
		Implements IOutputChannel

        Private remoteAddress_local As EndpointAddress

        Private via_local As Uri
        Private remoteEndPoint As EndPoint
        Private socket As Socket
        Private encoder As MessageEncoder
        Private parent As UdpChannelFactory

        Friend Sub New(ByVal factory As UdpChannelFactory, ByVal remoteAddress As EndpointAddress,
                       ByVal via As Uri, ByVal encoder As MessageEncoder)
            MyBase.New(factory)
            ' validate addressing arguments
            If Not String.Equals(via.Scheme, UdpConstants.Scheme, StringComparison.InvariantCultureIgnoreCase) Then
                Throw New ArgumentException(String.Format(CultureInfo.CurrentCulture,
                                                          "The scheme {0} specified in address is not supported.",
                                                          via.Scheme), "via")
            End If

            If via.IsDefaultPort Then
                Throw New ArgumentException(String.Format(CultureInfo.CurrentCulture,
                                                          "The address {0} must specify a remote port.", via), "via")
            End If

            ' convert the Uri host into an IP Address
            Dim remoteIP As IPAddress = Nothing
            Select Case via.HostNameType

                Case UriHostNameType.IPv4, UriHostNameType.IPv6
                    remoteIP = IPAddress.Parse(via.Host)

                Case UriHostNameType.Basic, UriHostNameType.Dns
                    Dim hostEntry As IPHostEntry = Dns.GetHostEntry(via.Host)
                    If hostEntry.AddressList.Length > 0 Then
                        remoteIP = hostEntry.AddressList(0)
                    Else
                        Throw New ArgumentException(String.Format(CultureInfo.CurrentCulture,
                                                                  "Failed to resolve remote host: {0}.",
                                                                  via.Host), "via")
                    End If
                Case Else
                    Throw New ArgumentException(String.Format(CultureInfo.CurrentCulture,
                                                              "Cannot determine the remote host address from {0}.",
                                                              Me.via_local.ToString()), "via")
                    Exit Select
            End Select

            If factory.Multicast AndAlso (Not UdpChannelHelpers.IsInMulticastRange(remoteIP)) Then
                Throw New ArgumentOutOfRangeException("remoteEndPoint", "Via must be in the valid multicast range.")
            End If

            Me.parent = factory
            Me.remoteAddress_local = remoteAddress
            Me.via_local = via
            Me.encoder = encoder
            Me.remoteEndPoint = New IPEndPoint(remoteIP, via.Port)
            Me.socket = New Socket(Me.remoteEndPoint.AddressFamily, SocketType.Dgram, ProtocolType.Udp)

            If parent.Multicast Then
                Me.socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, 1)
#If LATER Then ' Support outgoing interface
				If Me.remoteEndPoint.AddressFamily = AddressFamily.InterNetwork Then
					Me.socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastInterface, 
					                          IPAddress.HostToNetworkOrder(CInt(Fix(interfaceIndex))))
				Else ' we're IPv6
					Me.sendSocketV6.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.MulticastInterface, 
					                                CInt(Fix(interfaceIndex)))
				End If
#End If
            End If
        End Sub

#Region "IOutputChannel_Properties"
        Private ReadOnly Property RemoteAddress() As EndpointAddress Implements IOutputChannel.RemoteAddress
            Get
                Return Me.remoteAddress_local
            End Get
        End Property

        Private ReadOnly Property Via() As Uri Implements IOutputChannel.Via
            Get
                Return Me.via_local
            End Get
        End Property
#End Region

        Public ReadOnly Property Factory() As UdpChannelFactory
            Get
                Return Me.parent
            End Get
        End Property

      

        ''' <summary>
        ''' Open the channel for use. We don't have any blocking work to perform so this is a no-op
        ''' </summary>
        Protected Overrides Sub OnOpen(ByVal timeout As TimeSpan)
        End Sub

        Protected Overrides Function OnBeginOpen(ByVal timeout As TimeSpan, ByVal callback As AsyncCallback,
                                                 ByVal state As Object) As IAsyncResult
            Return New CompletedAsyncResult(callback, state)
        End Function

        Protected Overrides Sub OnEndOpen(ByVal result As IAsyncResult)
            CompletedAsyncResult.End(result)
        End Sub


#Region "Socket_Shutdown"
        ''' <summary>
        ''' Shutdown ungracefully
        ''' </summary>
        Protected Overrides Sub OnAbort()
            Me.socket.Close(0)
        End Sub

        ''' <summary>
        ''' Shutdown gracefully
        ''' </summary>
        Protected Overrides Sub OnClose(ByVal timeout As TimeSpan)
            Me.socket.Close()
        End Sub

        Protected Overrides Function OnBeginClose(ByVal timeout As TimeSpan, ByVal callback As AsyncCallback,
                                                  ByVal state As Object) As IAsyncResult
            Me.OnClose(timeout)
            Return New CompletedAsyncResult(callback, state)
        End Function

        Protected Overrides Sub OnEndClose(ByVal result As IAsyncResult)
            CompletedAsyncResult.End(result)
        End Sub
#End Region

#Region "Send_Synchronous"
        ''' <summary>
        ''' Address the Message and serialize it into a byte array.
        ''' </summary>
        Private Function EncodeMessage(ByVal message As Message) As ArraySegment(Of Byte)
            Try
                Me.remoteAddress_local.ApplyTo(message)
                Return encoder.WriteMessage(message, Integer.MaxValue, parent.BufferManager)
            Finally
                ' we've consumed the message by serializing it, so clean up.
                message.Close()
            End Try
        End Function

        Public Sub Send(ByVal message As Message) Implements IOutputChannel.Send
            If message Is Nothing Then
                Throw New ArgumentNullException("message")
            End If

            MyBase.ThrowIfDisposedOrNotOpen()

            'Obtain the transaction propagation token from the TransactionFlowProperty on the message.
            Dim txPropToken() As Byte = TransactionFlowProperty.Get(message)

            Dim messageBuffer As ArraySegment(Of Byte) = EncodeMessage(message)

            Dim txmsgBuffer() As Byte = TransactionMessageBuffer.WriteTransactionMessageBuffer(txPropToken, messageBuffer)
            If CLng(Fix(txmsgBuffer.Length)) > Me.Factory.MaxPacketSize Then
                Throw New CommunicationException("The output packet size is greater than the maximum size supported.")
            End If

            Try
                Dim bytesSent As Integer = Me.socket.SendTo(txmsgBuffer, 0, txmsgBuffer.Length,
                                                            SocketFlags.None, Me.remoteEndPoint)

                If bytesSent <> txmsgBuffer.Length Then
                    Throw New CommunicationException(String.Format(CultureInfo.CurrentCulture,
                                                                   "A Udp error occurred sending a message to {0}.",
                                                                   Me.remoteEndPoint))
                End If
            Catch socketException As SocketException
                Throw UdpChannelHelpers.ConvertTransferException(socketException)
            Finally
                ' we need to make sure buffers are always returned to the BufferManager.
                parent.BufferManager.ReturnBuffer(messageBuffer.Array)
            End Try
        End Sub

        Public Sub Send(ByVal message As Message, ByVal timeout As TimeSpan) Implements IOutputChannel.Send
            ' UDP doesn't block so we don't need timeouts.
            Me.Send(message)
        End Sub
#End Region

		#Region "Send_Asynchronous"
        Public Function BeginSend(ByVal message As Message, ByVal callback As AsyncCallback,
                                  ByVal state As Object) As IAsyncResult Implements IOutputChannel.BeginSend
            MyBase.ThrowIfDisposedOrNotOpen()
            Return New SendAsyncResult(Me, message, callback, state)
        End Function

        Public Function BeginSend(ByVal message As Message, ByVal timeout As TimeSpan, ByVal callback As AsyncCallback,
                                  ByVal state As Object) As IAsyncResult Implements IOutputChannel.BeginSend
            ' UDP doesn't block so we don't need timeouts.
            Return Me.BeginSend(message, callback, state)
        End Function

		Public Sub EndSend(ByVal result As IAsyncResult) Implements IOutputChannel.EndSend
			SendAsyncResult.End(result)
		End Sub

		''' <summary>
		''' Implementation of async send for Udp. 
		''' </summary>
		Private Class SendAsyncResult
			Inherits AsyncResult
			Private messageBuffer As ArraySegment(Of Byte)
			Private channel As UdpOutputChannel
			Private txmsgBuffer() As Byte

            Public Sub New(ByVal channel As UdpOutputChannel, ByVal message As Message,
                           ByVal callback As AsyncCallback, ByVal state As Object)
                MyBase.New(callback, state)
                Me.channel = channel

                'obtain the transaction propagation token from the TransactionFlowProperty on the message.
                Dim txPropToken() As Byte = TransactionFlowProperty.Get(message)

                Me.messageBuffer = channel.EncodeMessage(message)

                txmsgBuffer = TransactionMessageBuffer.WriteTransactionMessageBuffer(txPropToken, messageBuffer)
                If CLng(Fix(txmsgBuffer.Length)) > channel.Factory.MaxPacketSize Then
                    Throw New CommunicationException("The output packet size is greater than the maximum size supported.")
                End If

                Try
                    Dim result As IAsyncResult = Nothing
                    Try
                        result = channel.socket.BeginSendTo(txmsgBuffer, 0, txmsgBuffer.Length, SocketFlags.None,
                                                            channel.remoteEndPoint, New AsyncCallback(AddressOf OnSend), Me)
                    Catch socketException As SocketException
                        Throw UdpChannelHelpers.ConvertTransferException(socketException)
                    End Try

                    If Not result.CompletedSynchronously Then
                        Return
                    End If

                    CompleteSend(result, True)
                Catch
                    CleanupBuffer()
                    Throw
                End Try
            End Sub

			Private Sub CleanupBuffer()
				If messageBuffer.Array IsNot Nothing Then
					Me.channel.parent.BufferManager.ReturnBuffer(messageBuffer.Array)
					messageBuffer = New ArraySegment(Of Byte)()
				End If
			End Sub

			Private Sub CompleteSend(ByVal result As IAsyncResult, ByVal synchronous As Boolean)
				Try
					Dim bytesSent As Integer = channel.socket.EndSendTo(result)

					If bytesSent <> txmsgBuffer.Length Then
                        Throw New CommunicationException(String.Format(CultureInfo.CurrentCulture,
                                                                       "A Udp error occurred sending a message to {0}.",
                                                                       channel.remoteEndPoint))
					End If
				Catch socketException As SocketException
					Throw UdpChannelHelpers.ConvertTransferException(socketException)
				Finally
					CleanupBuffer()
				End Try

				MyBase.Complete(synchronous)
			End Sub

			Private Sub OnSend(ByVal result As IAsyncResult)
				If result.CompletedSynchronously Then
					Return
				End If

				Try
					CompleteSend(result, False)
				Catch e As Exception
					MyBase.Complete(False, e)
				End Try
			End Sub

            Public Overloads Shared Sub [End](ByVal result As IAsyncResult)
                AsyncResult.End(Of SendAsyncResult)(result)
            End Sub
		End Class
		#End Region
	End Class
End Namespace

File: framework/wcf/Extensibility/Transactions/TransactionMessagePropertyUdpTransport/CS/UdpTransport/UdpOutputChannel.cs
//  Copyright (c) Microsoft Corporation. All rights reserved.

using System;
using System.Diagnostics;
using System.Globalization;
using System.Net;
using System.Net.Sockets;
using System.ServiceModel;
using System.ServiceModel.Channels;

namespace Microsoft.ServiceModel.Samples
{
    /// <summary>
    /// IOutputChannel implementation for Udp.
    /// </summary>
    class UdpOutputChannel : ChannelBase, IOutputChannel
    {
        EndpointAddress remoteAddress;
        Uri via;
        EndPoint remoteEndPoint;
        Socket socket;
        MessageEncoder encoder;
        UdpChannelFactory parent;

        internal UdpOutputChannel(UdpChannelFactory factory, EndpointAddress remoteAddress, Uri via, MessageEncoder encoder)
            : base(factory)
        {
            // validate addressing arguments
            if (!string.Equals(via.Scheme, UdpConstants.Scheme, StringComparison.InvariantCultureIgnoreCase))
            {
                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, 
                    "The scheme {0} specified in address is not supported.", via.Scheme), "via");
            }

            if (via.IsDefaultPort)
            {
                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, 
                    "The address {0} must specify a remote port.", via), "via");
            }

            // convert the Uri host into an IP Address
            IPAddress remoteIP = null;
            switch (via.HostNameType)
            {
                default:
                    throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, 
                        "Cannot determine the remote host address from {0}.",
                            this.via.ToString()), "via");

                case UriHostNameType.IPv4:
                case UriHostNameType.IPv6:
                        remoteIP = IPAddress.Parse(via.Host);
                        break;

                case UriHostNameType.Basic:
                case UriHostNameType.Dns:
                    {
                        IPHostEntry hostEntry = Dns.GetHostEntry(via.Host);
                        if (hostEntry.AddressList.Length > 0)
                        {
                            remoteIP = hostEntry.AddressList[0];
                        }
                        else
                        {
                            throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, 
                                "Failed to resolve remote host: {0}.", via.Host), 
                                "via");
                        }
                        break;
                    }
            }

            if (factory.Multicast && !UdpChannelHelpers.IsInMulticastRange(remoteIP))
            {
                throw new ArgumentOutOfRangeException("remoteEndPoint", "Via must be in the valid multicast range.");
            }

            this.parent = factory;
            this.remoteAddress = remoteAddress;
            this.via = via;
            this.encoder = encoder;
            this.remoteEndPoint = new IPEndPoint(remoteIP, via.Port);
            this.socket = new Socket(this.remoteEndPoint.AddressFamily, SocketType.Dgram, ProtocolType.Udp);

            if (parent.Multicast)
            {
                this.socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, 1);
#if LATER // Support outgoing interface
                if (this.remoteEndPoint.AddressFamily == AddressFamily.InterNetwork)
                {
                    this.socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastInterface, IPAddress.HostToNetworkOrder((int)interfaceIndex));
                }
                else // we're IPv6
                {
                    this.sendSocketV6.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.MulticastInterface, (int)interfaceIndex);
                }
#endif
            }
        }

        #region IOutputChannel_Properties
        EndpointAddress IOutputChannel.RemoteAddress
        {
            get
            {
                return this.remoteAddress;
            }
        }

        Uri IOutputChannel.Via
        {
            get
            {
                return this.via;
            }
        }
        #endregion

        public UdpChannelFactory Factory
        {
            get
            {
                return this.parent;
            }
        }               

        public override T GetProperty<T>()
        {
            if (typeof(T) == typeof(IOutputChannel))
            {
                return (T)(object)this;
            }

            T messageEncoderProperty = this.encoder.GetProperty<T>();
            if (messageEncoderProperty != null)
            {
                return messageEncoderProperty;
            }

            return base.GetProperty<T>();
        }

        /// <summary>
        /// Open the channel for use. We don't have any blocking work to perform so this is a no-op
        /// </summary>
        protected override void OnOpen(TimeSpan timeout)
        {
        }

        protected override IAsyncResult OnBeginOpen(TimeSpan timeout, AsyncCallback callback, object state)
        {
            return new CompletedAsyncResult(callback, state);
        }

        protected override void OnEndOpen(IAsyncResult result)
        {
            CompletedAsyncResult.End(result);
        }


        #region Socket_Shutdown
        /// <summary>
        /// Shutdown ungracefully
        /// </summary>
        protected override void OnAbort()
        {
            this.socket.Close(0);
        }

        /// <summary>
        /// Shutdown gracefully
        /// </summary>
        protected override void OnClose(TimeSpan timeout)
        {
            this.socket.Close();
        }

        protected override IAsyncResult OnBeginClose(TimeSpan timeout, AsyncCallback callback, object state)
        {
            this.OnClose(timeout);
            return new CompletedAsyncResult(callback, state);
        }

        protected override void OnEndClose(IAsyncResult result)
        {
            CompletedAsyncResult.End(result);
        }
        #endregion

        #region Send_Synchronous
        /// <summary>
        /// Address the Message and serialize it into a byte array.
        /// </summary>
        ArraySegment<byte> EncodeMessage(Message message)
        {
            try
            {
                this.remoteAddress.ApplyTo(message);
                return encoder.WriteMessage(message, int.MaxValue, parent.BufferManager);
            }
            finally
            {
                // we've consumed the message by serializing it, so clean up
                message.Close();
            }
        }

        public void Send(Message message)
        {
            if (message == null)
                throw new ArgumentNullException("message");

            base.ThrowIfDisposedOrNotOpen();

            //obtain the transaction propagation token from the TransactionFlowProperty on the message
            byte[] txPropToken = TransactionFlowProperty.Get(message);
            
            ArraySegment<byte> messageBuffer = EncodeMessage(message);

            byte[] txmsgBuffer = TransactionMessageBuffer.WriteTransactionMessageBuffer(txPropToken, messageBuffer);
            if ((long)txmsgBuffer.Length > this.Factory.MaxPacketSize)
            {
                throw new CommunicationException("The output packet size is greater than the maximum size supported.");
            }

            try
            {                    
                int bytesSent = this.socket.SendTo(txmsgBuffer, 0, txmsgBuffer.Length,
                    SocketFlags.None, this.remoteEndPoint);

                if (bytesSent != txmsgBuffer.Length)
                {
                    throw new CommunicationException(string.Format(CultureInfo.CurrentCulture,
                        "A Udp error occurred sending a message to {0}.", this.remoteEndPoint));
                }               
            }
            catch (SocketException socketException)
            {
                throw UdpChannelHelpers.ConvertTransferException(socketException);
            }
            finally
            {
                // we need to make sure buffers are always returned to the BufferManager
                parent.BufferManager.ReturnBuffer(messageBuffer.Array);
            }
        }

        public void Send(Message message, TimeSpan timeout)
        {
            // UDP doesn't block so we don't need timeouts.
            this.Send(message);
        }
        #endregion

        #region Send_Asynchronous
        public IAsyncResult BeginSend(Message message, AsyncCallback callback, object state)
        {
            base.ThrowIfDisposedOrNotOpen();
            return new SendAsyncResult(this, message, callback, state);
        }

        public IAsyncResult BeginSend(Message message, TimeSpan timeout, AsyncCallback callback, object state)
        {
            // UDP doesn't block so we don't need timeouts.
            return this.BeginSend(message, callback, state);
        }

        public void EndSend(IAsyncResult result)
        {
            SendAsyncResult.End(result);
        }

        /// <summary>
        /// Implementation of async send for Udp. 
        /// </summary>
        class SendAsyncResult : AsyncResult
        {
            ArraySegment<byte> messageBuffer;
            UdpOutputChannel channel;
            byte[] txmsgBuffer;

            public SendAsyncResult(UdpOutputChannel channel, Message message, AsyncCallback callback, object state)
                : base(callback, state)
            {
                this.channel = channel;

                //obtain the transaction propagation token from the TransactionFlowProperty on the message
                byte[] txPropToken = TransactionFlowProperty.Get(message);

                this.messageBuffer = channel.EncodeMessage(message);

                txmsgBuffer = TransactionMessageBuffer.WriteTransactionMessageBuffer(txPropToken, messageBuffer);
                if ((long)txmsgBuffer.Length > channel.Factory.MaxPacketSize)
                {
                    throw new CommunicationException("The output packet size is greater than the maximum size supported.");
                }

                try
                {
                    IAsyncResult result = null;
                    try
                    {
                        result = channel.socket.BeginSendTo(txmsgBuffer, 0, txmsgBuffer.Length,
                            SocketFlags.None, channel.remoteEndPoint, new AsyncCallback(OnSend), this);
                    }
                    catch (SocketException socketException)
                    {
                        throw UdpChannelHelpers.ConvertTransferException(socketException);
                    }

                    if (!result.CompletedSynchronously)
                        return;

                    CompleteSend(result, true);
                }
                catch
                {
                    CleanupBuffer();
                    throw;
                }
            }

            void CleanupBuffer()
            {
                if (messageBuffer.Array != null)
                {
                    this.channel.parent.BufferManager.ReturnBuffer(messageBuffer.Array);
                    messageBuffer = new ArraySegment<byte>();
                }
            }

            void CompleteSend(IAsyncResult result, bool synchronous)
            {
                try
                {
                    int bytesSent = channel.socket.EndSendTo(result);

                    if (bytesSent != txmsgBuffer.Length)
                    {
                        throw new CommunicationException(string.Format(CultureInfo.CurrentCulture, 
                            "A Udp error occurred sending a message to {0}.", channel.remoteEndPoint));
                    }
                }
                catch (SocketException socketException)
                {
                    throw UdpChannelHelpers.ConvertTransferException(socketException);
                }
                finally
                {
                    CleanupBuffer();
                }

                base.Complete(synchronous);
            }

            void OnSend(IAsyncResult result)
            {
                if (result.CompletedSynchronously)
                    return;

                try
                {
                    CompleteSend(result, false);
                }
                catch (Exception e)
                {
                    base.Complete(false, e);
                }
            }

            public static void End(IAsyncResult result)
            {
                AsyncResult.End<SendAsyncResult>(result);
            }
        }
        #endregion
    }
}
